clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -119,33 +116,259 @@ void BoundsAnalysis::CollectBoundsVars(const Expr *E, DeclSetTy &BoundsVars) {
+    DeclSetTy BoundsVars;
+    CollectBoundsVars(BE, BoundsVars);
+    SuccEB->BoundsVars[V] = BoundsVars;
+  }
+}
+
+ExprPairTy BoundsAnalysis::SplitIntoBaseOffset(const Expr *E) {
+  // In order to make an expression uniform, we want to keep all DeclRefExprs
+  // on the LHS and all IntegerLiterals on the RHS.
+
+  if (IsDeclOperand(E))
+    return std::make_pair(GetDeclOperand(E), nullptr);
+
+  if (!isa<BinaryOperator>(E))
+    return std::make_pair(nullptr, nullptr);
+
+  const BinaryOperator *BO = dyn_cast<BinaryOperator>(E);
+  Expr *LHS = BO->getLHS();
+  Expr *RHS = BO->getRHS();
+
+  if (isa<ParenExpr>(LHS))
+    LHS = IgnoreCasts(LHS);
+  if (isa<ParenExpr>(RHS))
+    RHS = IgnoreCasts(RHS);
+
+  // Note: Assume p, q, r are DeclRefExprs and i, j are IntegerLiterals.
+
+  // Case 1: LHS is DeclRefExpr and RHS is IntegerLiteral. This expr is already
+  // uniform.
+  // p + i ==> return (p, i)
+  if (IsDeclOperand(LHS) && isa<IntegerLiteral>(RHS))
+    return std::make_pair(GetDeclOperand(LHS), RHS);
+
+  // Case 2: LHS is IntegerLiteral and RHS is DeclRefExpr. We simply need to
+  // swap LHS and RHS to make expr uniform.
+  // i + p ==> return (p, i)
+  if (isa<IntegerLiteral>(LHS) && IsDeclOperand(RHS))
+    return std::make_pair(GetDeclOperand(RHS), LHS);
+
+  // Case 3: LHS and RHS are both DeclRefExpr's. This means there is no
+  // IntegerLiteral in the expr. In this case, we return the incoming
+  // BinaryOperator expr with a nullptr for the RHS.
+  // p + q ==> return (p + q, nullptr)
+  if (IsDeclOperand(LHS) && IsDeclOperand(RHS))
+    return std::make_pair(BO, nullptr);
+
+  if (!isa<BinaryOperator>(LHS) &&
+      !isa<BinaryOperator>(RHS))
+    return std::make_pair(nullptr, nullptr);
+
+  // To make parsing simpler, we always try to keep BinaryOperator on the LHS.
+  if (isa<BinaryOperator>(RHS))
+    std::swap(LHS, RHS);
+
+  // If we reach here, the expr is one of these:
+  // Case 4: (p + q) + i
+  // Case 5: (p + j) + i
+  // Case 6: (p + q) + r
+  // Case 7: (p + j) + r
+  auto *BE = dyn_cast<BinaryOperator>(LHS);
+
+  // Recursively, make the LHS uniform.
+  ExprPairTy ExprPair = SplitIntoBaseOffset(BE);
+  const Expr *BinOpLHS = ExprPair.first;
+  const Expr *BinOpRHS = ExprPair.second;
+
+  // Expr is either Case 4 or Case 5 from above. ie: LHS is BinaryOperator
+  // and RHS is IntegerLiteral.
+  // (p + q) + i OR (p + j) + i
+  if (isa<IntegerLiteral>(RHS)) {
+
+    // Expr is Case 4. ie: The BinaryOperator expr does not have an
+    // IntegerLiteral on the RHS.
+    // (p + q) + i ==> return (p + q, i)
+    if (!BinOpRHS)
+      return std::make_pair(BE, RHS);
+
+    // Expr is Case 5. ie: The BinaryOperator expr has an IntegerLiteral on
+    // the RHS.
+    // (p + j) + i ==> return (p, j + i)
+    Expr::EvalResult R1, R2;
+    RHS->EvaluateAsInt(R1, Ctx);
+    BinOpRHS->EvaluateAsInt(R2, Ctx);
+    auto *I = new (Ctx)
+                IntegerLiteral(Ctx, R1.Val.getInt() + R2.Val.getInt(),
```

Comment:
You need to check for integer overflow or wraparound here.   I think it is fine to open an issue tracking this and place a TODO in the code.